src/core/tsi/ssl_transport_security.cc

```diff
@@ -1410,6 +1407,36 @@ static void ssl_handshaker_destroy(tsi_handshaker* self) {
   gpr_free(impl);
 }
 
+// Removes the bytes remaining in |impl->SSL|'s read BIO and writes them to
+// |bytes_remaining|.
+static tsi_result ssl_bytes_remaining(tsi_ssl_handshaker* impl,
```

Comment:
Not sure why we need to read unused bytes again. 
In ssl_handshaker_next, we already have received_bytes. We just need to call `BIO_pending(SSL_get_rbio(impl->ssl))` to find out how many bytes have been consumed and how many have not, then we can use the received_bytes buffer as unused_bytes (which is received_bytes + bytes_consumed).

Reply:
We have to remove the unused bytes from `impl->ssl`'s read BIO, otherwise it will try to decrypt the record twice. We could either:

1. Get the length (using `BIO_pending`) and clear `impl->ssl`'s read BIO, and then use received bytes.
2. Read the bytes from `impl->ssl` and set those to unused bytes.

There's no difference in memory use between the 2 options because `received_bytes` used to be copied. I opted for the second option because I think it is conceptually cleaner.

Reply:
I see. Thanks for explaining.
impl->network_io already consumes and buffers the unused bytes. Later impl->network_io and impl->ssl are handed over from handshaker to frame protector. Maybe in unprotect(), we can first process the unused bytes from BIO, while gRPC stack treats unused bytes size = 0. so that we can avoid BIO_read again and set unused bytes.

Reply:
This is what I tried originally but it created a problem (causing many tests to fail): if we set `unused_bytes_size` to zero, then the gRPC stack doesn't know it has any bytes waiting in the BIO that should be read. So, the server will just hang waiting for the client to send application data which the server already has, but doesn't know it has.

If we wanted to leave the bytes in BIO, then we would need to plumb a new way to indicate to the gRPC stack that it has bytes waiting to be read. But this is precisely what the unused bytes are already used for. This is why I think we should stick with this strategy.

Reply:
Make sense. I agree that we should stick to unused bytes then.